aequilibrae/project/network/link_types.py

```diff
@@ -0,0 +1,112 @@
+        default_link_type.save()
+
+        # We can also create a completely new link_type and add to the model
+        new_type = link_type('a')
+        new_type.link_type_name = 'Arterial'  # Only ASCII letters and *_* allowed
+        # other fields are not mandatory
+
+        # We then explicitly add it to the network
+        link_types.add(new_type)
+
+        # we can even keep editing and save it directly once we have added it to the project
+        new_type.lanes = 3
+        new_type.lane_capacity = 1100
+        new_type.save()
+    """
+
+    def __init__(self, net):
+        self.__all_types = []
+        self.conn = net.conn  # type: Connection
+        self.curr = net.conn.cursor()
+
+        self.__update_list_of_link_types()
+
+    def add(self, link_type: LinkType) -> None:
+        """ We add a new link type to the project"""
+        self.__update_list_of_link_types()
+        if link_type.link_type_id in self.__all_types:
+            raise ValueError("Link type already exists in the model")
+
+        self.curr.execute("insert into 'link_types'(link_type_id, link_type) Values(?,?)",
+                          [link_type.link_type_id, link_type.link_type])
+        self.conn.commit()
+        logger.info(f'Link Type {link_type.link_type}({link_type.link_type_id}) was added to the project')
+
+        link_type.save()
+        self.__update_list_of_link_types()
+
+    def drop(self, link_type_id: str) -> None:
+        """Removes the link_type with **link_type_id** from the project"""
+        try:
+            self.curr.execute(f'delete from link_types where link_type_id="{link_type_id}"')
+            self.conn.commit()
+        except IntegrityError as e:
+            logger.error(f'Failed to remove link_type {link_type_id}. {e.args}')
+            raise e
+        logger.warning(f'Link type {link_type_id} was successfully removed from the project database')
+        self.__update_list_of_link_types()
+
+    def get(self, link_type_id: str) -> LinkType:
+        """Get a link_type from the network by its **link_type_id**"""
+        self.__update_list_of_link_types()
+        if link_type_id not in self.__all_types:
+            raise ValueError(f'Link type {link_type_id} does not exist in the model')
+        return LinkType(link_type_id)
+
+    def get_by_name(self, link_type: str) -> LinkType:
+        """Get a mode from the network by its **mode_id**"""
+        self.__update_list_of_link_types()
+        self.curr.execute(f"select link_type_id from 'link_types' where link_type='{link_type}'")
+        found = self.curr.fetchone()
+        if len(found) == 0:
+            raise ValueError(f'Link type {link_type} does not exist in the model')
+        return LinkType(found[0])
+
+    def all_types(self) -> dict:
+        """Returns a dictionary with all mode objects available in the model. mode_id as key"""
+        self.__update_list_of_link_types()
+        return {x: LinkType(x) for x in self.__all_types}
```

Comment:
This will trigger a lot of DB queries that could all be done in one hit

Reply:
Do you see a way of doing it in one hit without duplicating the creation of LinkType objects?
I am also not too worried about this one, as its use will most likely be rare and the size of that table (link_types) should be incredibly small.

Reply:
I think the idea is what I mention in the comment above?

Reply:
Should be solved